File: starrysky/src/main/java/com/lzx/starrysky/playback/manager/PlaybackManager.java
/*
 * Copyright (C) 2014 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lzx.starrysky.playback.manager;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.ResultReceiver;
import android.os.SystemClock;
import android.support.annotation.NonNull;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import com.lzx.starrysky.BaseMediaInfo;
import com.lzx.starrysky.StarrySky;
import com.lzx.starrysky.ext.PlaybackStateCompatExt;
import com.lzx.starrysky.notification.INotification;
import com.lzx.starrysky.playback.player.ExoPlayback;
import com.lzx.starrysky.playback.player.Playback;
import com.lzx.starrysky.playback.queue.MediaQueue;
import com.lzx.starrysky.provider.MediaQueueProvider;
import com.lzx.starrysky.provider.MediaResource;
import com.lzx.starrysky.registry.ValidRegistry;
import com.lzx.starrysky.utils.delayaction.PlayValidManager;
import com.lzx.starrysky.utils.delayaction.Valid;


/**
 * 播放管理类
 */
public class PlaybackManager implements IPlaybackManager, Playback.Callback {

    private static final String TAG = "PlaybackManager";

    private MediaQueue mMediaQueue;
    private Playback mPlayback;
    private PlaybackServiceCallback mServiceCallback;
    private MediaSessionCallback mMediaSessionCallback;
    private INotification notification;
    private int currRepeatMode;
    private boolean shouldPlayNext = true; //是否可以播放下一首
    private boolean shouldPlayPre = true;  //是否可以播放上一首
    private PlaybackStateCompat.Builder stateBuilder;
    private Handler mHandler = new Handler(Looper.getMainLooper());


    public PlaybackManager(MediaQueue mediaQueue, Playback playback) {
        mMediaQueue = mediaQueue;
        mMediaSessionCallback = new MediaSessionCallback();
        mPlayback = playback;
        mPlayback.setCallback(this);
        currRepeatMode = PlaybackStateCompat.REPEAT_MODE_NONE;
    }

    @Override
    public void setServiceCallback(PlaybackServiceCallback serviceCallback) {
        mServiceCallback = serviceCallback;
    }

    @Override
    public void setMetadataUpdateListener(MediaQueueProvider.MetadataUpdateListener listener) {
        mMediaQueue.setMetadataUpdateListener(listener);
    }

    @Override
    public void registerNotification(INotification notification) {
        this.notification = notification;
    }

    public Playback getPlayback() {
        return mPlayback;
    }

    @Override
    public MediaSessionCompat.Callback getMediaSessionCallback() {
        return mMediaSessionCallback;
    }

    @Override
    public boolean isPlaying() {
        return mPlayback.isPlaying();
    }

    /**
     * 播放
     */
    @Override
    public void handlePlayRequest(boolean isPlayWhenReady) {
        ValidRegistry validRegistry = StarrySky.get().getRegistry().getValidRegistry();
        if (validRegistry.hasValid()) {
            PlayValidManager validManager = PlayValidManager.get();
            validManager.setAction(songInfo -> {
                BaseMediaInfo mediaInfo = mMediaQueue.songInfoToMediaInfo(songInfo);
                PlaybackManager.this.checkThreadHandPlayRequest(mediaInfo, isPlayWhenReady);
            });
            for (Valid valid : validRegistry.getValids()) {
                validManager.addValid(valid != null ? valid : new ValidRegistry.DefaultValid());
            }
            validManager.doCall(mMediaQueue.getCurrMediaInfo().getMediaId());
        } else {
            checkThreadHandPlayRequest(null, isPlayWhenReady);
        }
    }

    private void checkThreadHandPlayRequest(BaseMediaInfo mediaInfo, boolean isPlayWhenReady) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            mHandler.post(() -> handPlayRequestImpl(mediaInfo, isPlayWhenReady));
        } else {
            handPlayRequestImpl(mediaInfo, isPlayWhenReady);
        }
    }

    private void handPlayRequestImpl(BaseMediaInfo mediaInfo, boolean isPlayWhenReady) {
        MediaResource currentMusic = mMediaQueue.getCurrentMusic(mediaInfo);
        if (currentMusic != null) {
            if (isPlayWhenReady) {
                mServiceCallback.onPlaybackStart();
            }
            mPlayback.play(currentMusic, isPlayWhenReady);
        }
    }

    /**
     * 暂停
     */
    @Override
    public void handlePauseRequest() {
        if (mPlayback.isPlaying()) {
            mPlayback.pause();
            mServiceCallback.onPlaybackStop();
        }
    }

    /**
     * 停止
     */
    @Override
    public void handleStopRequest(String withError) {
        mPlayback.stop(true);
        mServiceCallback.onPlaybackStop();
        updatePlaybackState(false, withError);
    }

    /**
     * 快进
     */
    @Override
    public void handleFastForward() {
        mPlayback.onFastForward();
    }

    /**
     * 倒带
     */
    @Override
    public void handleRewind() {
        mPlayback.onRewind();
    }

    /**
     * 指定语速 refer 是否已当前速度为基数  multiple 倍率
     */
    @Override
    public void handleDerailleur(boolean refer, float multiple) {
        mPlayback.onDerailleur(refer, multiple);
    }

    /**
     * 更新播放状态
     */
    @Override
    public void updatePlaybackState(boolean isOnlyUpdateActions, String error) {
        if (isOnlyUpdateActions && stateBuilder != null) {
            //单独更新 Actions
            stateBuilder.setActions(getAvailableActions());
            mServiceCallback.onPlaybackStateUpdated(stateBuilder.build(), null);
        } else {
            long position = PlaybackStateCompat.PLAYBACK_POSITION_UNKNOWN;
            if (mPlayback != null && mPlayback.isConnected()) {
                position = mPlayback.getCurrentStreamPosition();
            }
            //构建一个播放状态对象
            stateBuilder = new PlaybackStateCompat.Builder()
                    .setActions(getAvailableActions());
            //获取播放器播放状态
            int state = mPlayback.getState();
            //如果错误信息不为 null 的时候，播放状态设为 STATE_ERROR
            if (error != null) {
                stateBuilder.setErrorMessage(error);
                state = PlaybackStateCompat.STATE_ERROR;
            }
            //设置播放状态
            stateBuilder.setState(state, position, 1.0f, SystemClock.elapsedRealtime());
            //设置当前活动的 songId
            MediaResource currentMusic = mMediaQueue.getCurrentMusic();
            MediaMetadataCompat currMetadata = null;
            if (currentMusic != null) {
                stateBuilder.setActiveQueueItemId(currentMusic.getQueueId());
                final String musicId = currentMusic.getMediaId();
                currMetadata = StarrySky.get().getMediaQueueProvider().getMediaMetadataCompatById(musicId);
            }
            //把状态回调出去
            mServiceCallback.onPlaybackStateUpdated(stateBuilder.build(), currMetadata);
            //如果是播放或者暂停的状态，更新一下通知栏
            if (state == PlaybackStateCompat.STATE_PLAYING || state == PlaybackStateCompat.STATE_PAUSED) {
                mServiceCallback.onNotificationRequired();
            }
        }
    }

    /**
     * 获取状态
     */
    private long getAvailableActions() {
        long actions = PlaybackStateCompat.ACTION_PLAY_PAUSE |
                PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID |
                PlaybackStateCompat.ACTION_PLAY_FROM_SEARCH |
                PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS |
                PlaybackStateCompat.ACTION_SKIP_TO_NEXT;
        if (mPlayback.isPlaying()) {
            actions |= PlaybackStateCompat.ACTION_PAUSE; //添加 ACTION_PAUSE
        } else {
            actions |= PlaybackStateCompat.ACTION_PLAY; //添加 ACTION_PLAY
        }
        if (!shouldPlayNext) {
            //在不能播放下一首的情况下，判断actions是否包含ACTION_SKIP_TO_NEXT，如果包含则清除
            if ((actions & PlaybackStateCompat.ACTION_SKIP_TO_NEXT) != 0) {
                actions &= ~PlaybackStateCompat.ACTION_SKIP_TO_NEXT;
            }
        } else {
            //判断 actions 是否包含 ACTION_SKIP_TO_NEXT，如果不包含，则添加
            if ((actions & PlaybackStateCompat.ACTION_SKIP_TO_NEXT) == 0) {
                actions |= PlaybackStateCompat.ACTION_SKIP_TO_NEXT;
            }
        }
        //同理
        if (!shouldPlayPre) {
            if ((actions & PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS) != 0) {
                actions &= ~PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS;
            }
        } else {
            if ((actions & PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS) == 0) {
                actions |= PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS;
            }
        }
        return actions;
    }

    /**
     * 播放器播放完成回调
     */
    @Override
    public void onCompletion() {
        updatePlaybackState(false, null);
        //单曲模式(播放当前就结束)
        if (currRepeatMode == PlaybackStateCompatExt.SINGLE_MODE_ONE) {
            return;
        }
        if (currRepeatMode == PlaybackStateCompat.REPEAT_MODE_NONE) {
            //顺序播放
            if (shouldPlayNext && mMediaQueue.skipQueuePosition(1)) {
                handlePlayRequest(true);
                mMediaQueue.updateMetadata();
            } else {
                handleStopRequest(null);
            }
        } else if (currRepeatMode == PlaybackStateCompat.REPEAT_MODE_ONE) {
            //单曲循环
            mPlayback.setCurrentMediaId("");
            handlePlayRequest(true);
        } else if (currRepeatMode == PlaybackStateCompat.REPEAT_MODE_ALL) {
            //列表循环
            if (shouldPlayNext && mMediaQueue.skipQueuePosition(1)) {
                handlePlayRequest(true);
                mMediaQueue.updateMetadata();
            } else {
                handleStopRequest(null);
            }
        }
    }

    /**
     * 播放器播放状态改变回调
     */
    @Override
    public void onPlaybackStatusChanged(int state) {
        updatePlaybackState(false, null);
    }

    /**
     * 播放器发送错误回调
     */
    @Override
    public void onError(String error) {
        updatePlaybackState(false, error);
    }

    /**
     * 设置当前播放 id
     */
    @Override
    public void setCurrentMediaId(String mediaId) {
        mMediaQueue.updateCurrPlayingMedia(mediaId);
    }

    /**
     * MusicManager API 方法的具体实现
     */
    private class MediaSessionCallback extends MediaSessionCompat.Callback {

        @Override
        public void onPrepare() {
            super.onPrepare();
            handlePlayRequest(false);
        }

        @Override
        public void onPrepareFromMediaId(String mediaId, Bundle extras) {
            super.onPrepareFromMediaId(mediaId, extras);
            mMediaQueue.updateCurrPlayingMedia(mediaId);
            handlePlayRequest(false);
        }

        @Override
        public void onPlay() {
            if (mMediaQueue.getCurrentMusic() == null) {
                return;
            }
            handlePlayRequest(true);
        }

        @Override
        public void onSkipToQueueItem(long queueId) {
            mMediaQueue.updateIndexByMediaId(String.valueOf(queueId));
            mMediaQueue.updateMetadata();
        }

        @Override
        public void onSeekTo(long position) {
            mPlayback.seekTo((int) position);
        }

        @Override
        public void onPlayFromMediaId(String mediaId, Bundle extras) {
            mMediaQueue.updateCurrPlayingMedia(mediaId);
            handlePlayRequest(true);
        }

        @Override
        public void onPause() {
            handlePauseRequest();
        }

        @Override
        public void onStop() {
            handleStopRequest(null);
        }

        @Override
        public void onSkipToNext() {
            if (shouldPlayNext) {
                if (mMediaQueue.skipQueuePosition(1)) {
                    handlePlayRequest(true);
                    mMediaQueue.updateMetadata();
                }
            }
        }

        @Override
        public void onSkipToPrevious() {
            if (shouldPlayPre) {
                if (mMediaQueue.skipQueuePosition(-1)) {
                    handlePlayRequest(true);
                    mMediaQueue.updateMetadata();
                }
            }
        }

        @Override
        public void onCustomAction(@NonNull String action, Bundle extras) {
            //updatePlaybackState(null);
        }

        @Override
        public void onFastForward() {
            super.onFastForward();
            handleFastForward();
        }

        @Override
        public void onRewind() {
            super.onRewind();
            handleRewind();
        }

        @Override
        public void onSetShuffleMode(int shuffleMode) {
            super.onSetShuffleMode(shuffleMode);
            if (shuffleMode == PlaybackStateCompat.SHUFFLE_MODE_NONE) {
                if (mPlayback != null) {
                    mMediaQueue.setNormalMode(mPlayback.getCurrentMediaId());
                }
            } else if (shuffleMode == PlaybackStateCompat.SHUFFLE_MODE_ALL) {
                mMediaQueue.setShuffledMode();
            }
            mServiceCallback.onShuffleModeUpdated(shuffleMode);
        }

        @Override
        public void onSetRepeatMode(int repeatMode) {
            super.onSetRepeatMode(repeatMode);
            currRepeatMode = repeatMode;
            mServiceCallback.onRepeatModeUpdated(repeatMode);
            updatePlaybackState(true, null);  //更新状态
        }

        /**
         * 自定义方法
         */
        @Override
        public void onCommand(String command, Bundle extras, ResultReceiver cb) {
            super.onCommand(command, extras, cb);
            if (command == null) {
                return;
            }
            if (INotification.ACTION_UPDATE_FAVORITE_UI.equals(command)) {
                boolean isFavorite = extras.getBoolean("isFavorite");
                if (notification != null) {
                    notification.updateFavoriteUI(isFavorite);
                }
            }
            if (INotification.ACTION_UPDATE_LYRICS_UI.equals(command)) {
                boolean isChecked = extras.getBoolean("isChecked");
                if (notification != null) {
                    notification.updateLyricsUI(isChecked);
                }
            }
            if (ExoPlayback.ACTION_CHANGE_VOLUME.equals(command)) {
                float audioVolume = extras.getFloat("AudioVolume");
                mPlayback.setVolume(audioVolume);
            }
            if (ExoPlayback.ACTION_DERAILLEUR.equals(command)) {
                boolean refer = extras.getBoolean("refer");
                float multiple = extras.getFloat("multiple");
                handleDerailleur(refer, multiple);
            }
        }
    }


}

File: starrysky/src/main/java/com/lzx/starrysky/playback/manager/PlaybackManager.kt
package com.lzx.starrysky.playback.manager

import android.os.Bundle
import android.os.ResultReceiver
import android.os.SystemClock
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import com.lzx.starrysky.StarrySky
import com.lzx.starrysky.control.RepeatMode
import com.lzx.starrysky.intercept.InterceptorCallback
import com.lzx.starrysky.intercept.InterceptorService
import com.lzx.starrysky.notification.INotification
import com.lzx.starrysky.playback.player.ExoPlayback
import com.lzx.starrysky.playback.player.Playback
import com.lzx.starrysky.playback.queue.MediaQueue
import com.lzx.starrysky.provider.IMediaSourceProvider
import com.lzx.starrysky.provider.SongInfo
import com.lzx.starrysky.utils.MainLooper
import com.lzx.starrysky.utils.StarrySkyUtils

class PlaybackManager constructor(
    private val mediaQueue: MediaQueue, private val playback: Playback
) : IPlaybackManager, Playback.Callback {

    private var mServiceCallback: IPlaybackManager.PlaybackServiceCallback? = null
    private val mMediaSessionCallback: MediaSessionCallback
    private var notification: INotification? = null
    private var stateBuilder: PlaybackStateCompat.Builder? = null
    private val interceptorService: InterceptorService = InterceptorService()

    init {
        mMediaSessionCallback = MediaSessionCallback()
        playback.setCallback(this)
    }

    override val mediaSessionCallback: MediaSessionCompat.Callback
        get() = mMediaSessionCallback

    override val isPlaying: Boolean
        get() = playback.isPlaying

    override fun setServiceCallback(serviceCallback: IPlaybackManager.PlaybackServiceCallback) {
        mServiceCallback = serviceCallback
    }

    override fun setMetadataUpdateListener(listener: IMediaSourceProvider.MetadataUpdateListener) {
        mediaQueue.setMetadataUpdateListener(listener)
    }

    /**
     * 主动触发的都从正常的队列中取
     * 非主动触发的根据播放模式在不同的队列中取
     */
    override fun handlePlayRequest(isPlayWhenReady: Boolean, isActiveTrigger: Boolean) {
        val playSongInfo = mediaQueue.getCurrentSongInfo(isActiveTrigger) //要播放的歌曲信息
        interceptorService.doInterceptions(playSongInfo, object : InterceptorCallback {
            override fun onContinue(songInfo: SongInfo?) {
                MainLooper.instance.runOnUiThread(Runnable {
                    mediaQueue.updateMediaMetadata(playSongInfo) //更新媒体封面信息
                    handPlayRequestImpl(songInfo, isPlayWhenReady)
                })
            }

            override fun onInterrupt(exception: Throwable?) {
                MainLooper.instance.runOnUiThread(Runnable {
                    updatePlaybackState(playSongInfo, isOnlyUpdateActions = false, isError = true,
                        error = exception?.message)
                    playback.currentMediaId = ""
                })
            }
        })
    }

    private fun handPlayRequestImpl(songInfo: SongInfo?, isPlayWhenReady: Boolean) {
        songInfo?.let {
            if (isPlayWhenReady) {
                mServiceCallback?.onPlaybackStart()
            }
            playback.play(it, isPlayWhenReady)
        }
    }

    override fun handlePauseRequest() {
        if (playback.isPlaying) {
            playback.pause()
            mServiceCallback?.onPlaybackStop(false)
        }
    }

    override fun handleStopRequest(withError: String?) {
        playback.stop()
        mServiceCallback?.onPlaybackStop(true)
        val hasError = withError?.isNotEmpty() ?: false
        updatePlaybackState(playback.currPlayInfo, false, hasError, withError)
    }

    override fun handleFastForward() {
        playback.onFastForward()
    }

    override fun handleRewind() {
        playback.onRewind()
    }

    override fun handleDerailleur(refer: Boolean, multiple: Float) {
        playback.onDerailleur(refer, multiple)
    }

    override fun updatePlaybackState(
        currPlayInfo: SongInfo?,
        isOnlyUpdateActions: Boolean,
        isError: Boolean,
        error: String?
    ) {
        if (isOnlyUpdateActions) {
            //单独更新 Actions
            stateBuilder?.setActions(getAvailableActions())
            mServiceCallback?.onPlaybackStateUpdated(stateBuilder?.build(), null)
        } else {
            //当前播放进度
            var currentStreamPosition = PlaybackStateCompat.PLAYBACK_POSITION_UNKNOWN
            if (playback.isConnected) {
                currentStreamPosition = playback.currentStreamPosition
            }
            //构建一个播放状态对象
            stateBuilder = PlaybackStateCompat.Builder().setActions(getAvailableActions())
            //获取播放器播放状态
            var state = playback.playbackState
            //发生错误的时候，播放状态设为 STATE_ERROR
            if (isError) {
                stateBuilder?.setErrorMessage(if (!error.isNullOrEmpty()) error else "错误信息为 null")
                state = Playback.STATE_ERROR
            }
            //设置播放状态
            stateBuilder?.setState(state, currentStreamPosition, 1.0f,
                SystemClock.elapsedRealtime())

            //获取当前的 MediaMetadataCompat
            var currMetadata: MediaMetadataCompat? = null
            currPlayInfo?.let {
                stateBuilder?.setActiveQueueItemId(-1L)
                currMetadata =
                    StarrySky.get().mediaQueueProvider().getMediaMetadataById(it.songId)
            }
            //把状态回调出去
            val playbackState = stateBuilder?.build()
            mServiceCallback?.onPlaybackStateUpdated(playbackState, currMetadata)
            //如果是播放或者暂停的状态，更新一下通知栏
            if (state == Playback.STATE_PLAYING || state == Playback.STATE_PAUSED) {
                notification?.startNotification(currPlayInfo, playbackState)
            }
        }
    }

    /**
     * 获取状态
     */
    private fun getAvailableActions(): Long {
        var actions = PlaybackStateCompat.ACTION_PLAY_PAUSE or
            PlaybackStateCompat.ACTION_PLAY_FROM_MEDIA_ID or
            PlaybackStateCompat.ACTION_PLAY_FROM_SEARCH or
            PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS or
            PlaybackStateCompat.ACTION_SKIP_TO_NEXT
        actions = if (playback.isPlaying) {
            actions or PlaybackStateCompat.ACTION_PAUSE //添加 ACTION_PAUSE
        } else {
            actions or PlaybackStateCompat.ACTION_PLAY //添加 ACTION_PLAY
        }
        var shouldPlayNext = true
        var shouldPlayPre = true
        val repeatMode = StarrySkyUtils.repeatMode
        if (repeatMode.repeatMode != RepeatMode.REPEAT_MODE_SHUFFLE) {
            //如果没开启循环并且当前歌曲是最后一首，则不能下一首
            shouldPlayNext = !(!repeatMode.isLoop && mediaQueue.currSongIsLastSong())
            shouldPlayPre = !(!repeatMode.isLoop && mediaQueue.currSongIsFirstSong())
        }
        if (!shouldPlayNext) {
            //在不能播放下一首的情况下，判断actions是否包含ACTION_SKIP_TO_NEXT，如果包含则清除
            if (actions and PlaybackStateCompat.ACTION_SKIP_TO_NEXT != 0L) {
                actions = actions and PlaybackStateCompat.ACTION_SKIP_TO_NEXT.inv()
            }
        } else {
            //判断 actions 是否包含 ACTION_SKIP_TO_NEXT，如果不包含，则添加
            if (actions and PlaybackStateCompat.ACTION_SKIP_TO_NEXT == 0L) {
                actions = actions or PlaybackStateCompat.ACTION_SKIP_TO_NEXT
            }
        }
        //同理
        if (!shouldPlayPre) {
            if (actions and PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS != 0L) {
                actions = actions and PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS.inv()
            }
        } else {
            if (actions and PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS == 0L) {
                actions = actions or PlaybackStateCompat.ACTION_SKIP_TO_PREVIOUS
            }
        }
        return actions
    }

    override fun registerNotification(notification: INotification?) {
        this.notification = notification
    }

    /**
     * 播放器回调播放完成时会回调这里
     */
    override fun onPlaybackCompletion() {
        updatePlaybackState(playback.currPlayInfo, isOnlyUpdateActions = false, isError = false,
            error = null)
        val repeatMode = StarrySkyUtils.repeatMode
        when (repeatMode.repeatMode) {
            //顺序播放
            RepeatMode.REPEAT_MODE_NONE -> {
                if (repeatMode.isLoop) {
                    skipToNextSongImpl(1)
                } else {
                    if (!mediaQueue.currSongIsLastSong()) {
                        skipToNextSongImpl(1)
                    }
                }
            }
            //单曲播放
            RepeatMode.REPEAT_MODE_ONE -> {
                playback.currentMediaId = ""
                if (repeatMode.isLoop) {
                    handlePlayRequest(isPlayWhenReady = true, isActiveTrigger = false)
                }
            }
            //随机播放
            RepeatMode.REPEAT_MODE_SHUFFLE -> {
                skipToNextSongImpl(1)
            }
            //倒序播放
            RepeatMode.REPEAT_MODE_REVERSE -> {
                if (repeatMode.isLoop) {
                    skipToNextSongImpl(-1)
                } else {
                    if (!mediaQueue.currSongIsFirstSong()) {
                        skipToNextSongImpl(-1)
                    }
                }
            }
        }
    }

    /**
     * 播放器回调状态改变时回调这里
     */
    override fun onPlaybackStatusChanged(songInfo: SongInfo?, state: Int) {
        updatePlaybackState(songInfo, isOnlyUpdateActions = false, isError = false, error = null)
    }

    /**
     * 播放器回调错误时回调这里
     */
    override fun onPlaybackError(songInfo: SongInfo?, error: String) {
        updatePlaybackState(songInfo, isOnlyUpdateActions = false, isError = true, error = error)
    }

    private fun skipToNextSongImpl(amount: Int) {
        if (mediaQueue.skipQueuePosition(amount)) {
            handlePlayRequest(isPlayWhenReady = true, isActiveTrigger = false)
        }
    }

    /**
     * MusicManager API 方法的具体实现
     */
    inner class MediaSessionCallback : MediaSessionCompat.Callback() {

        /**
         * 缓冲不播放
         */
        override fun onPrepare() {
            super.onPrepare()
            handlePlayRequest(isPlayWhenReady = false, isActiveTrigger = true)
        }

        /**
         * 根据 id 缓冲
         */
        override fun onPrepareFromMediaId(mediaId: String?, extras: Bundle?) {
            super.onPrepareFromMediaId(mediaId, extras)
            mediaId?.apply {
                mediaQueue.updateIndexBySongId(this)
                handlePlayRequest(isPlayWhenReady = false, isActiveTrigger = true)
            }
        }

        /**
         * 根据 id 播放
         */
        override fun onPlayFromMediaId(mediaId: String?, extras: Bundle?) {
            super.onPlayFromMediaId(mediaId, extras)
            mediaId?.apply {
                mediaQueue.updateIndexBySongId(this)
                handlePlayRequest(isPlayWhenReady = true, isActiveTrigger = true)
            }
        }

        /**
         * 暂停后播放
         */
        override fun onPlay() {
            super.onPlay()
            playback.currPlayInfo?.let {
                handlePlayRequest(isPlayWhenReady = true, isActiveTrigger = true)
            }
        }

        /**
         * 暂停
         */
        override fun onPause() {
            super.onPause()
            handlePauseRequest()
        }

        /**
         * 停止
         */
        override fun onStop() {
            super.onStop()
            handleStopRequest(null)
        }

        /**
         * 拖进度条
         */
        override fun onSeekTo(pos: Long) {
            super.onSeekTo(pos)
            playback.seekTo(pos)
            if (playback.playbackState == Playback.STATE_PAUSED) {
                onPlay()
            }
        }

        /**
         * 下一首
         */
        override fun onSkipToNext() {
            super.onSkipToNext()
            if (mediaQueue.skipQueuePosition(1)) {
                handlePlayRequest(isPlayWhenReady = true, isActiveTrigger = true)
            }
        }

        /**
         * 上一首
         */
        override fun onSkipToPrevious() {
            super.onSkipToPrevious()
            if (mediaQueue.skipQueuePosition(-1)) {
                handlePlayRequest(isPlayWhenReady = true, isActiveTrigger = true)
            }
        }

        /**
         * 快进
         */
        override fun onFastForward() {
            super.onFastForward()
            handleFastForward()
        }

        /**
         * 快退
         */
        override fun onRewind() {
            super.onRewind()
            handleRewind()
        }

        /**
         * 自定义命令
         */
        override fun onCommand(command: String?, extras: Bundle?, cb: ResultReceiver?) {
            super.onCommand(command, extras, cb)
            if (command == null) {
                return
            }
            when (command) {
                ExoPlayback.ACTION_CHANGE_VOLUME -> {
                    val audioVolume = extras?.getFloat("AudioVolume") ?: 0F
                    playback.volume = audioVolume
                }
                ExoPlayback.ACTION_DERAILLEUR -> {
                    val refer = extras?.getBoolean("refer") ?: false
                    val multiple = extras?.getFloat("multiple") ?: 0F
                    handleDerailleur(refer, multiple)
                }
                //播放模式
                RepeatMode.KEY_REPEAT_MODE -> {
                    val repeatMode = StarrySkyUtils.repeatMode.repeatMode
                    if (repeatMode == RepeatMode.REPEAT_MODE_SHUFFLE) {
                        StarrySky.get().mediaQueueProvider().updateShuffleSongList()
                    } else {
                        mediaQueue.updateIndexBySongId(playback.currentMediaId)
                    }
                    updatePlaybackState(null, isOnlyUpdateActions = true, isError = false, error = null)
                }
                else -> {
                    //通知栏相关
                    notification?.onCommand(command, extras)
                }
            }
        }
    }
}